frontend/model/contracts/group.js

```diff
@@ -1124,6 +1118,34 @@ sbp('chelonia/defineContract', {
       }
       await sbp('gi.db/archive/save', key, proposals)
       sbp('okTurtles.events/emit', PROPOSAL_ARCHIVED, [proposalHash, proposal])
+    },
+    'gi.contracts/group/archivePayments': async function (contractID, archivingPayments) {
+      const { paymentsByPeriod, payments } = archivingPayments
+      const { username } = sbp('state/vuex/state').loggedIn
+
+      for (const period of Object.keys(paymentsByPeriod).sort()) {
+        const archPaymentsByPeriodKey = `paymentsByPeriod/${username}/${contractID}`
+        const archPaymentsByPeriod = await sbp('gi.db/archive/load', archPaymentsByPeriodKey) || {}
+        const archPaymentsKey = `payments/${username}/${contractID}`
+        let archPayments = await sbp('gi.db/archive/load', archPaymentsKey) || {}
+
+        archPaymentsByPeriod[period] = paymentsByPeriod[period]
+        archPayments = merge(archPayments, payments)
+
+        while (Object.keys(archPaymentsByPeriod).length > MAX_ARCHIVED_PERIODS) {
+          const shouldBeDeletedPeriod = Object.keys(archPaymentsByPeriod).sort().shift()
+          const paymentHashes = paymentHashesFromPaymentPeriod(archPaymentsByPeriod[shouldBeDeletedPeriod])
+
+          for (const hash of paymentHashes) {
+            delete archPayments[hash]
+          }
+          delete archPaymentsByPeriod[shouldBeDeletedPeriod]
+        }
+
+        await sbp('gi.db/archive/save', archPaymentsByPeriodKey, archPaymentsByPeriod)
+        await sbp('gi.db/archive/save', archPaymentsKey, archPayments)
+      }
+      sbp('okTurtles.events/emit', PAYMENTS_ARCHIVED, { paymentsByPeriod, payments })
```

Comment:
@Silver-IT So this PR seems to be well written and ready to merge, however I have one remaining doubt about this approach that I mentioned on Slack (feel free to reply there or here), and that is:

- On slower and older devices, it might still be quite a lot to save and load all of the payments payments in one go.
- The whole point of archiving the payments in the first place is to not have to load all of them into memory. Yet with this PR we will still end up doing that.

Although storing the array of archived payments using pagination<sup>1</sup>  is more complicated...

So, what do you think, we could merge this now, and then work on archiving using pages in a future PR?


<sup>1</sup> e.g. by storing the length of the array and paginating the keys themselves, for example, by adding the periodstamp to the key name, or by adding numbers to the keyname.

Reply:
Yeah, it's really good point. I am 100% sure that we should archive payments using pagination.
I will create an issue for this and wanna implement in the future PR.

Reply:
Created [this issue](https://github.com/okTurtles/group-income/issues/1448).